plugins/vpp/ifplugin/vppcalls/dump_interface_vppcalls.go

```diff
@@ -153,14 +165,28 @@ func (h *IfVppHandler) DumpInterfaces() (map[uint32]*InterfaceDetails, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to dump unnumbered interfaces: %v", err)
 	}
-	// Get vrf for every interface and fill DHCP if set
+	// Get interface VRF for every IP family, fill DHCP if set and resolve unnumbered interface setup
 	for _, ifData := range ifs {
-		// VRF
-		vrf, err := h.GetInterfaceVrf(ifData.Meta.SwIfIndex)
+		// VRF is stored in metadata for both, IPv4 and IPv6. If the interface is an IPv6 interface (it contains at least
+		// one IPv6 address), appropriate VRF is stored also in modelled data
+		ipv4Vrf, err := h.GetInterfaceVrf(ifData.Meta.SwIfIndex)
 		if err != nil {
 			return nil, fmt.Errorf("interface dump: failed to get VRF from interface %d: %v", ifData.Meta.SwIfIndex, err)
 		}
-		ifData.Interface.Vrf = vrf
+		ifData.Meta.VrfIPv4 = ipv4Vrf
+		ipv6Vrf, err := h.GetInterfaceVrfIPv6(ifData.Meta.SwIfIndex)
+		if err != nil {
+			return nil, fmt.Errorf("interface dump: failed to get IPv6 VRF from interface %d: %v", ifData.Meta.SwIfIndex, err)
+		}
+		ifData.Meta.VrfIPv6 = ipv6Vrf
+		if isIPv6If, err := h.isIpv6Interface(ifData.Interface); err != nil {
+			return ifs, err
+		} else if isIPv6If {
+			ifData.Interface.Vrf = ipv6Vrf
```

Comment:
The case when interface has both IPv4 and IPv6 addresses and vrfs do not match is not handled (I would just set very large number to ifData.Interface.Vrf and thus enforce interface re-create) .
In the refactor I especially focus on resync and making sure that agent always gets back in-sync from whatever state - we cannot even assume that configuration is in a consistent state from the agent perspective, because agent may crash and restart in-between calls to SetInterfaceVrf for IPv4 and IPv6.

Reply:
If VRF does not match, the interface will be re-created anyway, there is no need to enforce it. Since we can dump two VRF values from the VPP, but store only one in the proto-modeled configuration, the method should ensure that the correct value is in the model (which does not define IP version) and specifies it in the metadata.
With both IPv4 and IPv6, they will be in same-number VRF (even if VPP treats it as two separate), so if we store IPv6 VRF, that the correct value.
The dump method just returns the VPP state, so I'm not sure whether I understand why some large number should be stored in VRF. If we want to enforce re-create, resync can do it itself. The method also serves as RPC output, I think putting any made-up numbers may be confusing.